### PR TITLE
Add step to wait for opensearch

### PR DESCRIPTION
**Pull request by iHiD**
Try and fix this!

Screenshot 2023-08-21 at 15 02 46

### CODE REVIEW

**Comment by iHiD** [comment]
Conclusion is that opensearch sometimes just never becomes ready. We could ideally conditionally find a way to restart the job in this situation. @ErikSchierboom Any ideas?

**Comment by ErikSchierboom** [comment]
My initial though is to detect that opensearch isn't running and then to re-trigger the workflow via a `workflow_dispatch` message. https://www.eliostruyf.com/restart-github-actions-workflow-failed/

File: bin/wait_for_opensearch [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+timeout="10s"
+status="yellow"
+url="'${OPENSEARCH_HOST}/_cluster/health?wait_for_status=${status}&timeout=${timeout}'"
```
Comment (ErikSchierboom) [inline bin/wait_for_opensearch]: We can use opensearch's built-in functionality for this: https://opensearch.org/docs/latest/api-reference/cluster-api/cluster-health/

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -229,6 +228,21 @@ jobs:
...
+      ###
+      # Wait for everything to be ready
+      - name: Wait for open search to become ready
+        id: wait-for-opensearch
+        env:
+          OPENSEARCH_HOST: https://127.0.0.1:${{ job.services.opensearch.ports['9200'] }}
+        run: ./bin/wait_for_opensearch
+
+      - name: Re-run job if open-search failed to become ready
+        if: ${{ failure() }}
```
Comment (ErikSchierboom) [inline .github/workflows/tests.yml]: This means it only runs when the previous step failed.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -229,6 +228,21 @@ jobs:
...
+        env:
+          GH_TOKEN: ${{ secrets.RERUN_WORKFLOWS_PAT }}
```
Comment (ErikSchierboom) [inline .github/workflows/tests.yml]: We need a separate token to be able to re-run workflow runs

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -229,6 +228,21 @@ jobs:
...
+        run: |
+          gh run rerun --failed ${{ github.run_id }}
```
Comment (ErikSchierboom) [inline .github/workflows/tests.yml]: This should only re-run the current run, not the entire workflow

File: .github/workflows/rerun-failed.yml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+name: Rerun failed
+
+on:
+  workflow_run:
+    workflows: [Run]
+    types: [completed]
```
Comment (ErikSchierboom) [inline .github/workflows/rerun-failed.yml]: This ensures that we can safely restart the workflow, as it fires when the entire run completes (could be in a failed state)

File: .github/workflows/rerun-failed.yml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  on-failure:
+    runs-on: ubuntu-22.04
+    if: ${{ github.event.workflow_run.conclusion == 'failure' }}
+    steps:
+      - name: Check if OpenSearch failed to start
+        id: check-opensearch
+        run: |
+          jobs=$(gh api ${{ github.event.workflow_run.jobs_url }})
+          failed=$(echo "${jobs}" | jq '.jobs | any(.steps | any(.name == "Wait for open search to become ready" and .conclusion == "failure"))')
+          echo "failed=${failed}" >> "$GITHUB_OUTPUT"
```
Comment (ErikSchierboom) [inline .github/workflows/rerun-failed.yml:20]: We don't just want to restart any run, but only those where opensearch failed to start

File: .github/workflows/rerun-failed.yml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        env:
+          GH_TOKEN: ${{ github.token }}
+
+      - name: Rerun job
+        if: ${{ steps.check-opensearch.outputs.failed == 'true' }}
+        run: |
+          gh api --method POST ${{ github.event.workflow_run.rerun_url }}
+        env:
+          GITHUB_TOKEN: ${{ secrets.WORKFLOWS_PAT }}
```
Comment (ErikSchierboom) [inline .github/workflows/rerun-failed.yml:29]: A fine-grained token worked! Wouldn't you know

File: .github/workflows/tests.yml [hunk 8]
```diff
@@ -229,6 +228,13 @@ jobs:
           bundle exec rails r bin/monitor-manifest
           bundle exec rails test:prepare
 
+      ###
+      # Wait for everything to be ready
+      - name: Wait for open search to become ready
+        env:
+          OPENSEARCH_HOST: https://127.0.0.1:${{ job.services.opensearch.ports['9200'] }}
+        run: ./bin/wait_for_opensearch
```
Comment (ErikSchierboom) [inline .github/workflows/tests.yml:236]: I wonder at what step we should be having this check. We're not doing a bunch of other things first that will be meaningless if opensearch won't start

File: .github/workflows/rerun-failed.yml (hunk 7 above)
Comment (iHiD) [reply .github/workflows/rerun-failed.yml:29]: Wow!

**Pull request merged**